<a href="https://colab.research.google.com/github/furrutiav/NLP-competition-2/blob/main/test_stopwords_competencia2_ner_felipe_urrutia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Instalamos torchtext que nos facilitará la vida en el pre-procesamiento del formato ConLL.
!pip install -U torchtext==0.10.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 7.6 MB 17.6 MB/s 
     |████████████████████████████████| 831.4 MB 2.8 kB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.11.0+cu113
    Uninstalling torch-1.11.0+cu113:
      Successfully uninstalled torch-1.11.0+cu113
  Attempting uninstall: torchtext
    Found existing installation: torchtext 0.12.0
    Uninstalling torchtext-0.12.0:
      Successfully uninstalled torchtext-0.12.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchvision 0.12.0+cu113 requires torch==1.11.0, but you have torch 1.9.0 which is incompatible.
torchaudio 0.11.0+cu113 requires torch==1.11.0, but you have torch 1.9.0 which is incompatible.


In [ ]:
import torch
from torchtext import data, datasets, legacy


# Garantizar reproducibilidad de los experimentos
SEED = 1234
torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

Descargamos los datos de entrenamiento, validación y prueba en nuestro directorio de trabajo

In [ ]:
#%%capture

!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/train.txt -nc # Dataset de Entrenamiento
!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/dev.txt -nc    # Dataset de Validación (Para probar y ajustar el modelo)
!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/test.txt -nc  # Dataset de la Competencia. Estos datos solo contienen los tokens. ¡¡SON LOS QUE DEBEN SER PREDICHOS!!

--2022-06-26 19:47:31--  https://github.com/dccuchile/CC6205/releases/download/v1.0/train.txt
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/196273020/77198f00-c145-11eb-83d1-11e647241ab6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20220626%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220626T194731Z&X-Amz-Expires=300&X-Amz-Signature=25b43a27837ad698b6cfe29233d5f0251fe7194e2576fdc7f6f9837fad42a79c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=196273020&response-content-disposition=attachment%3B%20filename%3Dtrain.txt&response-content-type=application%2Foctet-stream [following]
--2022-06-26 19:47:31--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/196273020/77198f00-c145-11eb-83d1-11e647241ab6?X-Amz-Algorithm=AWS4-H

In [ ]:
import torch
from torchtext import data, datasets, legacy
import pandas as pd
import numpy as np
# import nltk
# nltk.download('stopwords')
# from nltk.corpus import stopwords
import string
import time

In [ ]:
!pip install -U spacy
import spacy
!python -m spacy download es_core_news_md
nlp = spacy.load('es_core_news_md')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 42.3 MB 1.1 MB/s 
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')


In [ ]:
with open('train.txt') as f:
    lines_train = f.readlines()

with open('dev.txt') as f:
    lines_dev = f.readlines()

In [ ]:
def preprocessing(lines, method="masked", verbose=True):
  _times = []
  lines_stopwords = []
  N = len(lines)
  for i, line in enumerate(lines):
    start = time.time()
    new_line = line
    stop = False
    if len(line.split())==2:
      w, tag = line.split()
      doc = nlp(w)
      if doc[0].is_stop:
        stop = True
        w = "<stopword>"
        new_line = f"{w} {tag}\n"
    if method == "masked": lines_stopwords.append(new_line)
    else:
      if not stop: lines_stopwords.append(new_line)
    
    _times.append(time.time()-start)
    if ((i+1)%1000==0 or (i+1)==len(lines)) and verbose:
      end_expected = (N-(i+1))*np.mean(_times)
      end_expected_hrs = end_expected//(60*60)
      end_expected_min = (end_expected-end_expected_hrs*(60*60))//60
      end_expected_sec = end_expected - end_expected_hrs*(60*60) - end_expected_min*60
          
      trans_expected = np.sum(_times)
      trans_expected_hrs = trans_expected//(60*60)
      trans_expected_min = (trans_expected-trans_expected_hrs*(60*60))//60
      trans_expected_sec = trans_expected - trans_expected_hrs*(60*60) - trans_expected_min*60

      print(f"""{(i+1)}/{N}, progress: {100*(i+1)/N: .1f} %, dt: {_times[-1]: .2f}, exp. dt: {np.mean(_times): .2f} p/m {np.std(_times): .2f} s, t. trans: {trans_expected_hrs: .1f} hrs {trans_expected_min: .1f} min {trans_expected_sec: .1f} s, exp. t. end: {end_expected_hrs: .1f} hrs {end_expected_min: .1f} min {end_expected_sec: .1f} s""")

  return lines_stopwords

In [ ]:
lines_stopwords_masked = preprocessing(lines_train, method="masked", verbose=True)
lines_stopwords_masked      

1000/157246, progress:  0.6 %, dt:  0.02, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  6.1 s, exp. t. end:  0.0 hrs  15.0 min  59.2 s
2000/157246, progress:  1.3 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  12.7 s, exp. t. end:  0.0 hrs  16.0 min  25.6 s
3000/157246, progress:  1.9 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  17.0 s, exp. t. end:  0.0 hrs  14.0 min  33.7 s
4000/157246, progress:  2.5 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  21.2 s, exp. t. end:  0.0 hrs  13.0 min  31.9 s
5000/157246, progress:  3.2 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  25.5 s, exp. t. end:  0.0 hrs  12.0 min  56.8 s
6000/157246, progress:  3.8 %, dt:  0.00, exp. dt:  0.00 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  29.8 s, exp. t. end:  0.0 hrs  12.0 min  30.2 s
7000/157246, progress:  4.5 %, dt:  0.00, exp. dt:  0.00 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  33.9 s, exp. t. end:

['K08 B-Disease\n',
 '. I-Disease\n',
 '1 I-Disease\n',
 '- O\n',
 'PERDIDA B-Disease\n',
 '<stopword> I-Disease\n',
 'DIENTES I-Disease\n',
 'DEBIDA I-Disease\n',
 '<stopword> I-Disease\n',
 'ACCIDENTE I-Disease\n',
 ', I-Disease\n',
 'EXTRACCION I-Disease\n',
 '<stopword> I-Disease\n',
 'ENF I-Disease\n',
 '. I-Disease\n',
 'PERIODONTAL I-Disease\n',
 'LOCAL I-Disease\n',
 '/ O\n',
 '<stopword> O\n',
 'solicita O\n',
 'Protesis B-Procedure\n',
 'Parcial I-Procedure\n',
 'superior I-Procedure\n',
 '<stopword> I-Procedure\n',
 'inferior I-Procedure\n',
 '\n',
 '- O\n',
 'PERIODONTITIS B-Disease\n',
 'CRÓNICA I-Disease\n',
 'MODERADA I-Disease\n',
 '/ O\n',
 '- O\n',
 'Fundamento O\n',
 'Clínico O\n',
 'APS O\n',
 ': O\n',
 'PACIENTE O\n',
 '53 O\n',
 'AÑOS O\n',
 '<stopword> O\n',
 'ANTECEDENTES O\n',
 'MEDICOS O\n',
 ', O\n',
 'PRESENTA O\n',
 'PERIODONTITIS B-Disease\n',
 'MARGINAL I-Disease\n',
 'CRONICA I-Disease\n',
 'MODERADA I-Disease\n',
 'GENERALIZADA I-Disease\n',
 ', O\n',
 

In [ ]:
lines_stopwords_wo = [line for line in lines_stopwords_masked if len(line.split())==2 and line.split()[0] != "<stopword>"]
lines_stopwords_wo

['K08 B-Disease\n',
 '. I-Disease\n',
 '1 I-Disease\n',
 '- O\n',
 'PERDIDA B-Disease\n',
 'DIENTES I-Disease\n',
 'DEBIDA I-Disease\n',
 'ACCIDENTE I-Disease\n',
 ', I-Disease\n',
 'EXTRACCION I-Disease\n',
 'ENF I-Disease\n',
 '. I-Disease\n',
 'PERIODONTAL I-Disease\n',
 'LOCAL I-Disease\n',
 '/ O\n',
 'solicita O\n',
 'Protesis B-Procedure\n',
 'Parcial I-Procedure\n',
 'superior I-Procedure\n',
 'inferior I-Procedure\n',
 '- O\n',
 'PERIODONTITIS B-Disease\n',
 'CRÓNICA I-Disease\n',
 'MODERADA I-Disease\n',
 '/ O\n',
 '- O\n',
 'Fundamento O\n',
 'Clínico O\n',
 'APS O\n',
 ': O\n',
 'PACIENTE O\n',
 '53 O\n',
 'AÑOS O\n',
 'ANTECEDENTES O\n',
 'MEDICOS O\n',
 ', O\n',
 'PRESENTA O\n',
 'PERIODONTITIS B-Disease\n',
 'MARGINAL I-Disease\n',
 'CRONICA I-Disease\n',
 'MODERADA I-Disease\n',
 'GENERALIZADA I-Disease\n',
 ', O\n',
 'PRESENTO O\n',
 'ABSCESO B-Disease\n',
 'PERIODONTAL I-Disease\n',
 'PZA I-Disease\n',
 '7 I-Disease\n',
 '- I-Disease\n',
 '8 I-Disease\n',
 ', O\n',
 'D

In [ ]:
lines_dev_stopwords_masked = preprocessing(lines_dev, method="masked", verbose=True)
lines_dev_stopwords_masked 

1000/17467, progress:  5.7 %, dt:  0.01, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  5.7 s, exp. t. end:  0.0 hrs  1.0 min  33.2 s
2000/17467, progress:  11.5 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  12.7 s, exp. t. end:  0.0 hrs  1.0 min  38.3 s
3000/17467, progress:  17.2 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  17.1 s, exp. t. end:  0.0 hrs  1.0 min  22.3 s
4000/17467, progress:  22.9 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  21.5 s, exp. t. end:  0.0 hrs  1.0 min  12.2 s
5000/17467, progress:  28.6 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  25.8 s, exp. t. end:  0.0 hrs  1.0 min  4.3 s
6000/17467, progress:  34.4 %, dt:  0.00, exp. dt:  0.01 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  30.1 s, exp. t. end:  0.0 hrs  0.0 min  57.6 s
7000/17467, progress:  40.1 %, dt:  0.00, exp. dt:  0.00 p/m  0.00 s, t. trans:  0.0 hrs  0.0 min  34.4 s, exp. t. end:  0.0 hr

['POSIBILIDAD O\n',
 '<stopword> O\n',
 '<stopword> O\n',
 'ENDODONCIA B-Procedure\n',
 '<stopword> O\n',
 'PIEZA B-Body_Part\n',
 '2 I-Body_Part\n',
 '. I-Body_Part\n',
 '1 I-Body_Part\n',
 '( O\n',
 '<stopword> B-Disease\n',
 '. I-Disease\n',
 'D I-Disease\n',
 '. I-Disease\n',
 '<stopword> I-Disease\n',
 ') O\n',
 'PIEZA B-Body_Part\n',
 'TREPANADA B-Procedure\n',
 '<stopword> O\n',
 'PACIENTE O\n',
 'TRATADO O\n',
 '<stopword> O\n',
 'ANTIBIOTICOS B-Medication\n',
 '<stopword> O\n',
 'ANALGESICOS B-Medication\n',
 '. O\n',
 '\n',
 'Paciente O\n',
 '<stopword> O\n',
 '77 O\n',
 'años O\n',
 'desdentada B-Disease\n',
 'totalmente I-Disease\n',
 '<stopword> I-Disease\n',
 'párate I-Disease\n',
 'superior I-Disease\n',
 '<stopword> I-Disease\n',
 'inferior I-Disease\n',
 'portadora O\n',
 '<stopword> O\n',
 'prótesis O\n',
 '<stopword> O\n',
 '<stopword> O\n',
 '<stopword> O\n',
 '<stopword> O\n',
 '\n',
 '- O\n',
 'DIABETES B-Disease\n',
 'MELLITUS I-Disease\n',
 'INSULINOREQUIRIENTE 

In [ ]:
with open('train_wo.txt', 'a') as f:
    f.writelines(lines_stopwords_wo)
with open('train_masked.txt', 'a') as f:
    f.writelines(lines_stopwords_masked)
with open('dev_masked.txt', 'a') as f:
    f.writelines(lines_dev_stopwords_masked)

In [ ]:
# Primer Field: TEXT. Representan los tokens de la secuencia
TEXT = legacy.data.Field(lower=False) 
TEXT_masked = legacy.data.Field(lower=False) 
TEXT_wo = legacy.data.Field(lower=False) 

# Segundo Field: NER_TAGS. Representan los Tags asociados a cada palabra.
NER_TAGS = legacy.data.Field(unk_token=None)
fields = (("text", TEXT), ("nertags", NER_TAGS))

NER_TAGS_masked = legacy.data.Field(unk_token=None)
fields_masked = (("text", TEXT_masked), ("nertags", NER_TAGS_masked))

NER_TAGS_wo = legacy.data.Field(unk_token=None)
fields_wo = (("text", TEXT_wo), ("nertags", NER_TAGS_wo))


La idea es que usando los fields que definimos antes, le indiquemos a la clase cómo cargar los datasets de prueba, validación y test.

In [ ]:
train_data, valid_data, test_data = legacy.datasets.SequenceTaggingDataset.splits(
    path="./",
    train="train.txt",
    validation="dev.txt",
    test="test.txt",
    fields=fields,
    encoding="utf-8",
    separator=" "
)

train_data_masked, valid_data_masked, test_data_masked = legacy.datasets.SequenceTaggingDataset.splits(
    path="./",
    train="train_masked.txt",
    validation="dev_masked.txt",
    test="test.txt",
    fields=fields_masked,
    encoding="utf-8",
    separator=" "
)

train_data_wo, valid_data_wo, test_data_wo = legacy.datasets.SequenceTaggingDataset.splits(
    path="./",
    train="train_wo.txt",
    validation="dev.txt",
    test="test.txt",
    fields=fields_wo,
    encoding="utf-8",
    separator=" "
)

In [ ]:
TEXT.build_vocab(train_data)
NER_TAGS.build_vocab(train_data)

TEXT_masked.build_vocab(train_data_masked)
NER_TAGS_masked.build_vocab(train_data_masked)

TEXT_wo.build_vocab(train_data_wo)
NER_TAGS_wo.build_vocab(train_data_wo)

In [ ]:
print(f"Tokens únicos en TEXT: {len(TEXT.vocab)}")
print(f"Tokens únicos en NER_TAGS: {len(NER_TAGS.vocab)}")

print(f"Tokens únicos en TEXT_masked: {len(TEXT_masked.vocab)}")
print(f"Tokens únicos en NER_TAGS_masked: {len(NER_TAGS_masked.vocab)}")

print(f"Tokens únicos en TEXT_wo: {len(TEXT_wo.vocab)}")
print(f"Tokens únicos en NER_TAGS_wo: {len(NER_TAGS_wo.vocab)}")

Tokens únicos en TEXT: 17591
Tokens únicos en NER_TAGS: 12
Tokens únicos en TEXT_masked: 16968
Tokens únicos en NER_TAGS_masked: 12
Tokens únicos en TEXT_wo: 16967
Tokens únicos en NER_TAGS_wo: 12


In [ ]:
#Veamos las posibles etiquetas que hemos cargado:
print(NER_TAGS.vocab.itos)

print(NER_TAGS_masked.vocab.itos)

print(NER_TAGS_wo.vocab.itos)

['<pad>', 'O', 'I-Disease', 'B-Disease', 'I-Body_Part', 'B-Body_Part', 'B-Procedure', 'I-Procedure', 'B-Medication', 'B-Family_Member', 'I-Medication', 'I-Family_Member']
['<pad>', 'O', 'I-Disease', 'B-Disease', 'I-Body_Part', 'B-Body_Part', 'B-Procedure', 'I-Procedure', 'B-Medication', 'B-Family_Member', 'I-Medication', 'I-Family_Member']
['<pad>', 'O', 'I-Disease', 'B-Disease', 'I-Body_Part', 'B-Body_Part', 'B-Procedure', 'I-Procedure', 'B-Medication', 'B-Family_Member', 'I-Medication', 'I-Family_Member']


Observen que ademas de los tags NER, tenemos \<pad\>, el cual es generado por el dataloader para cumplir con el padding de cada oración.

Veamos ahora los tokens mas frecuentes y especiales:

In [ ]:
# Seteamos algunas variables que nos serán de utilidad mas adelante...
UNK_IDX = TEXT.vocab.stoi[TEXT.unk_token]
PAD_IDX = TEXT.vocab.stoi[TEXT.pad_token]

PAD_TAG_IDX = NER_TAGS.vocab.stoi[NER_TAGS.pad_token]
O_TAG_IDX = NER_TAGS.vocab.stoi['O']

In [ ]:
BATCH_SIZE = 16  # disminuir si hay problemas de ram.

# Usar cuda si es que está disponible.
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using', device)

# Dividir datos entre entrenamiento y test. Si van a hacer algún sort no puede ser sobre
# el conjunto de testing ya que al hacer sus predicciones sobre el conjunto de test sin etiquetas
# debe conservar el orden original para ser comparado con los golden_labels. 

train_iterator, valid_iterator, test_iterator = legacy.data.BucketIterator.splits(
    (train_data, valid_data, test_data),
    batch_size=BATCH_SIZE,
    device=device,
    sort=False,
)

train_iterator_masked, valid_iterator_masked, test_iterator_masked = legacy.data.BucketIterator.splits(
    (train_data_masked, valid_data_masked, test_data_masked),
    batch_size=BATCH_SIZE,
    device=device,
    sort=False,
)

train_iterator_wo, valid_iterator_wo, test_iterator_wo = legacy.data.BucketIterator.splits(
    (train_data_wo, valid_data_wo, test_data_wo),
    batch_size=BATCH_SIZE,
    device=device,
    sort=False,
)

Using cuda


In [ ]:
!pip install seqeval

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 43 kB 2.5 MB/s 
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16180 sha256=8ebbeacdfdf2a3674c91a707f1bcbb030953e7ab528cc4ae43d5c125352b74cd
  Stored in directory: /root/.cache/pip/wheels/05/96/ee/7cac4e74f3b19e3158dce26a20a1c86b3533c43ec72a549fd7
Successfully built seqeval


In [ ]:
# Definimos las métricas

from seqeval.metrics import f1_score, precision_score, recall_score

def calculate_metrics(preds, y_true, pad_idx=0):
    """
    Calcula precision, recall y f1 de cada batch.
    """

    # Obtener el indice de la clase con probabilidad mayor. (clases)
    y_pred = preds.argmax(dim=1, keepdim=True)

    # filtramos <pad> para calcular los scores.
    mask = [(y_true != pad_idx)]
    y_pred = y_pred[mask]
    y_true = y_true[mask]

    # traemos a la cpu
    y_pred = y_pred.view(-1).to('cpu').numpy()
    y_true = y_true.to('cpu').numpy()
    y_pred = [[NER_TAGS.vocab.itos[v] for v in y_pred]]
    y_true = [[NER_TAGS.vocab.itos[v] for v in y_true]]
    
    # calcular scores
    f1 = f1_score(y_true, y_pred, mode='strict')
    precision = precision_score(y_true, y_pred, mode='strict')
    recall = recall_score(y_true, y_pred, mode='strict')

    return precision, recall, f1

In [147]:
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim


# Definir la red
class NER_RNN(nn.Module):
    def __init__(self, 
                 input_dim, 
                 embedding_dim, 
                 hidden_dim, 
                 output_dim,
                 n_layers, 
                 bidirectional, 
                 dropout, 
                 pad_idx):

        super().__init__()

        # Capa de embedding
        self.embedding = nn.Embedding(input_dim,
                                      embedding_dim,
                                      padding_idx=pad_idx,
                                      )

        # Capa LSTM
        self.lstm = nn.LSTM(embedding_dim,
                           hidden_dim,
                           num_layers=n_layers,
                           bidirectional=bidirectional, 
                           dropout = dropout if n_layers > 1 else 0)

        # Capa de salida
        self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim,
                            output_dim)

        # Dropout
        self.dropout = nn.Dropout(dropout)

    def forward(self, text):

        #text = [sent len, batch size]

        # Convertir lo enviado a embedding
        embedded = self.dropout(self.embedding(text))
        
        outputs, (hidden, cell) = self.lstm(embedded)
        #embedded = [sent len, batch size, emb dim]

        # Pasar los embeddings por la rnn (LSTM)

        #output = [sent len, batch size, hid dim * n directions]
        #hidden/cell = [n layers * n directions, batch size, hid dim]

        # Predecir usando la capa de salida.
        predictions = self.fc(self.dropout(outputs))
        #predictions = [sent len, batch size, output dim]

        return predictions

#### **Hiperparámetros de la red**



In [148]:
# tamaño del vocabulario. recuerden que la entrada son vectores bag of word(one-hot).
EMBEDDING_DIM = 200  # dimensión de los embeddings.
HIDDEN_DIM = 128  # dimensión de la capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab)  # número de clases

N_LAYERS = 3  # número de capas.
DROPOUT = 0.5
BIDIRECTIONAL = False

# Creamos nuestro modelo.
baseline_model = NER_RNN(len(TEXT.vocab), EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX)

baseline_model_name = 'baseline'  

# Creamos nuestro modelo.
masked_model = NER_RNN(len(TEXT_masked.vocab), EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX)

masked_model_name = 'masked'  

# Creamos nuestro modelo.
wo_model = NER_RNN(len(TEXT_wo.vocab), EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX)

wo_model_name = 'wo'  

In [149]:
baseline_n_epochs = 10

#### Definimos la función de loss

In [152]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.stoi[NER_TAGS.pad_token]
baseline_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)
masked_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)
wo_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)

In [153]:
model_1 = baseline_model
model_name_1 = baseline_model_name
n_epochs_1 = baseline_n_epochs
criterion_1 = baseline_criterion

---------------

### Modelo 2

In [154]:
model_2 = masked_model
model_name_2 = masked_model_name
n_epochs_2 = baseline_n_epochs
criterion_2 = masked_criterion

---------------


### Modelo 3

In [155]:
model_3 = wo_model
model_name_3 = wo_model_name
n_epochs_3 = baseline_n_epochs
criterion_3= wo_criterion

In [157]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

def train(model, iterator, optimizer, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.train()

    # Por cada batch del iterador de la época:
    for batch in iterator:

        # Extraemos el texto y los tags del batch que estamos procesado
        text = batch.text
        tags = batch.nertags

        # Reiniciamos los gradientes calculados en la iteración anterior
        optimizer.zero_grad()

        #text = [sent len, batch size]

        # Predecimos los tags del texto del batch.
        predictions = model(text)

        #predictions = [sent len, batch size, output dim]
        #tags = [sent len, batch size]

        # Reordenamos los datos para calcular la loss
        predictions = predictions.view(-1, predictions.shape[-1])
        tags = tags.view(-1)

        #predictions = [sent len * batch size, output dim]



        # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
        loss = criterion(predictions, tags)
        
        # Calculamos el accuracy
        precision, recall, f1 = calculate_metrics(predictions, tags)

        # Calculamos los gradientes
        loss.backward()

        # Actualizamos los parámetros de la red
        optimizer.step()

        # Actualizamos el loss y las métricas
        epoch_loss += loss.item()
        epoch_precision += precision
        epoch_recall += recall
        epoch_f1 += f1

    return epoch_loss / len(iterator), epoch_precision / len(
        iterator), epoch_recall / len(iterator), epoch_f1 / len(iterator)

def evaluate(model, iterator, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.eval()

    # Indicamos que ahora no guardaremos los gradientes
    with torch.no_grad():
        # Por cada batch
        for batch in iterator:

            text = batch.text
            tags = batch.nertags

            # Predecimos
            predictions = model(text)

            predictions = predictions.view(-1, predictions.shape[-1])
            tags = tags.view(-1)

            # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
            loss = criterion(predictions, tags)

            # Calculamos las métricas
            precision, recall, f1 = calculate_metrics(predictions, tags)

            # Actualizamos el loss y las métricas
            epoch_loss += loss.item()
            epoch_precision += precision
            epoch_recall += recall
            epoch_f1 += f1

    return epoch_loss / len(iterator), epoch_precision / len(
        iterator), epoch_recall / len(iterator), epoch_f1 / len(iterator)

### Modelo seleccionado

In [164]:
for index_model in range(3):
  model = [model_1, model_2, model_3][index_model]
  model_name = [model_name_1, model_name_2, model_name_3][index_model]
  criterion = [criterion_1, criterion_2, criterion_3][index_model]
  n_epochs = [n_epochs_1, n_epochs_2, n_epochs_3][index_model]

  train_iterator_ = [train_iterator, train_iterator_masked, train_iterator_wo][index_model]
  valid_iterator_ = [valid_iterator, valid_iterator_masked, valid_iterator_wo][index_model]


  def init_weights(m):
      # Inicializamos los pesos como aleatorios
      for name, param in m.named_parameters():
          nn.init.normal_(param.data, mean=0, std=0.1) 
          
      # Seteamos como 0 los embeddings de UNK y PAD.
      model.embedding.weight.data[UNK_IDX] = torch.zeros(EMBEDDING_DIM)
      model.embedding.weight.data[PAD_IDX] = torch.zeros(EMBEDDING_DIM)

  model.apply(init_weights)

  print(f'[{model_name}] El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

  # Optimizador
  optimizer = optim.Adam(model.parameters())

  # Enviamos el modelo y la loss a cuda (en el caso en que esté disponible)
  model = model.to(device)
  criterion = criterion.to(device)

  best_valid_loss = float('inf')

  for epoch in range(n_epochs):

      start_time = time.time()

      # Recuerdo: train_iterator y valid_iterator contienen el dataset dividido en batches.

      # Entrenar
      train_loss, train_precision, train_recall, train_f1 = train(
          model, train_iterator_, optimizer, criterion)

      # Evaluar (valid = validación)
      valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
          model, valid_iterator_, criterion)

      end_time = time.time()

      epoch_mins, epoch_secs = epoch_time(start_time, end_time)

      # Si obtuvimos mejores resultados, guardamos este modelo en el almacenamiento (para poder cargarlo luego)
      # Si detienen el entrenamiento prematuramente, pueden cargar el modelo en el siguiente recuadro de código.
      if valid_loss < best_valid_loss:
          best_valid_loss = valid_loss
          torch.save(model.state_dict(), '{}.pt'.format(model_name))
      # Si ya no mejoramos el loss de validación, terminamos de entrenar.

      print(f'[{model_name}] Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
      print(
          f'[{model_name}] \tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
      )
      print(
          f'[{model_name}] \t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
      )

  # cargar el mejor modelo entrenado.
  model.load_state_dict(torch.load('{}.pt'.format(model_name)))

  # Limpiar ram de cuda
  torch.cuda.empty_cache()

  valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
      model, valid_iterator_, criterion)

  print(
      f'[{model_name}] Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} | Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
  )


[baseline] El modelo actual tiene 3,952,900 parámetros entrenables.


/usr/local/lib/python3.7/dist-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: <pad> seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/usr/local/lib/python3.7/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


[baseline] Epoch: 01 | Epoch Time: 0m 7s
[baseline] 	Train Loss: 0.928 | Train f1: 0.30 | Train precision: 0.46 | Train recall: 0.23
[baseline] 	 Val. Loss: 0.575 |  Val. f1: 0.58 |  Val. precision: 0.72 | Val. recall: 0.49
[baseline] Epoch: 02 | Epoch Time: 0m 6s
[baseline] 	Train Loss: 0.526 | Train f1: 0.62 | Train precision: 0.70 | Train recall: 0.57
[baseline] 	 Val. Loss: 0.452 |  Val. f1: 0.66 |  Val. precision: 0.73 | Val. recall: 0.61
[baseline] Epoch: 03 | Epoch Time: 0m 6s
[baseline] 	Train Loss: 0.395 | Train f1: 0.72 | Train precision: 0.75 | Train recall: 0.69
[baseline] 	 Val. Loss: 0.427 |  Val. f1: 0.69 |  Val. precision: 0.74 | Val. recall: 0.65
[baseline] Epoch: 04 | Epoch Time: 0m 6s
[baseline] 	Train Loss: 0.322 | Train f1: 0.76 | Train precision: 0.78 | Train recall: 0.75
[baseline] 	 Val. Loss: 0.415 |  Val. f1: 0.71 |  Val. precision: 0.74 | Val. recall: 0.69
[baseline] Epoch: 05 | Epoch Time: 0m 6s
[baseline] 	Train Loss: 0.272 | Train f1: 0.80 | Train precisio

Normal:

Val. Loss: 0.398 |  Val. f1: 0.71 | Val. precision: 0.73 | Val. recall: 0.70

Masked:

Val. Loss: 0.495 |  Val. f1: 0.67 | Val. precision: 0.65 | Val. recall: 0.70

Without:

Val. Loss: 0.489 |  Val. f1: 0.67 | Val. precision: 0.68 | Val. recall: 0.67